File: go/helpers/classification.go
// SPDX-License-Identifier: MIT

package helpers

import (
	vstar "hop.top/vstar"
	"hop.top/vstar/hashing"
)

// classProp is the wire name for the CLASS property
// (RFC 5545 §3.8.1.3). Applies to VEVENT, VTODO and VJOURNAL.
const classProp = "CLASS"

// transpProp is the wire name for the TRANSP property
// (RFC 5545 §3.8.2.7). Applies to VEVENT only.
const transpProp = "TRANSP"

// validClass reports whether v is one of the three wire values
// RFC 5545 §3.8.1.3 enumerates for CLASS.
func validClass(v vstar.Class) bool {
	switch v {
	case vstar.ClassPublic, vstar.ClassPrivate, vstar.ClassConfidential:
		return true
	default:
		return false
	}
}

// classApplies reports whether ct is a component type that admits a
// CLASS property per RFC 5545 §3.8.1.3.
func classApplies(ct vstar.CompType) bool {
	switch ct {
	case vstar.CompEvent, vstar.CompTodo, vstar.CompJournal:
		return true
	default:
		return false
	}
}

// validTransp reports whether v is one of the two wire values
// RFC 5545 §3.8.2.7 enumerates for TRANSP.
func validTransp(v vstar.Transp) bool {
	switch v {
	case vstar.TranspOpaque, vstar.TranspTransparent:
		return true
	default:
		return false
	}
}

// Class returns the parsed CLASS value of c when present and valid.
// Returns ok=false when CLASS is absent or carries an unrecognized
// value.
//
// An absent CLASS reports ok=false rather than (ClassPublic, true),
// even though RFC 5545 §3.8.1.3 assigns PUBLIC by default. The
// getter reports what is on the wire; the default is a separate,
// opt-in question answered by ClassOrDefault. Two reasons:
//
//   - "(value, bool)" means "present and recognized" everywhere else
//     in this package (Status, EventStatus, Transp, …). Having one
//     getter alone return ok=true for an absent property would make
//     the flag untrustworthy.
//   - Callers that must distinguish "explicitly PUBLIC" from
//     "unset" — round-trip fidelity, diffing, supersession
//     projection — cannot recover that distinction once the getter
//     has folded it away, whereas callers that only want the
//     effective value get it from ClassOrDefault at zero cost.
//
// This mirrors how the package already treats the other defaulted
// property it models: RelatedTo applies the RFC's "PARENT" default
// for an omitted RELTYPE, but does so in a function that returns
// the effective value with no ok flag to contradict.
func Class(c vstar.Component) (vstar.Class, bool) {
	p, ok := c.Get(classProp)
	if !ok {
		return "", false
	}
	v := vstar.Class(p.Value)
	if !validClass(v) {
		return "", false
	}
	return v, true
}

// ClassOrDefault returns the effective CLASS of c, applying the
// RFC 5545 §3.8.1.3 default of PUBLIC when the property is absent
// or carries an unrecognized value.
//
// Unrecognized values fall back to the default rather than being
// surfaced: the RFC treats an unknown CLASS token as equivalent to
// PRIVATE only for IANA/X- registrations it cannot resolve, and V*
// does not model those. Flagging the value is the validate
// package's job (VS044's sibling rules), not this accessor's.
func ClassOrDefault(c vstar.Component) vstar.Class {
	if v, ok := Class(c); ok {
		return v
	}
	return vstar.ClassPublic
}

// SetClass writes the CLASS property on c and refreshes the
// X-VSTAR-HASH last. No-op when c is nil, c.Type does not admit
// CLASS (VEVENT, VTODO, VJOURNAL per RFC 5545 §3.8.1.3), or v is
// not one of the three valid Class constants.
func SetClass(c *vstar.Component, v vstar.Class) {
	if c == nil || !classApplies(c.Type) || !validClass(v) {
		return
	}
	c.Set(vstar.Property{Name: classProp, Value: string(v)})
	hashing.SetXVSTAR(c)
}

// Transp returns the parsed TRANSP value of c when present and
// valid. Returns ok=false when TRANSP is absent or carries an
// unrecognized value. See Class for why an absent property is not
// reported as the RFC default here; use TranspOrDefault for the
// effective value.
func Transp(c vstar.Component) (vstar.Transp, bool) {
	p, ok := c.Get(transpProp)
	if !ok {
		return "", false
	}
	v := vstar.Transp(p.Value)
	if !validTransp(v) {
		return "", false
	}
	return v, true
}

// TranspOrDefault returns the effective TRANSP of c, applying the
// RFC 5545 §3.8.2.7 default of OPAQUE when the property is absent
// or carries an unrecognized value. OPAQUE means the event
// consumes free/busy time.
func TranspOrDefault(c vstar.Component) vstar.Transp {
	if v, ok := Transp(c); ok {
		return v
	}
	return vstar.TranspOpaque
}

// SetTransp writes the TRANSP property on c and refreshes the
// X-VSTAR-HASH last. No-op when c is nil, c.Type is not CompEvent
// (TRANSP is VEVENT-only per RFC 5545 §3.8.2.7), or v is not one of
// the two valid Transp constants.
func SetTransp(c *vstar.Component, v vstar.Transp) {
	if c == nil || c.Type != vstar.CompEvent || !validTransp(v) {
		return
	}
	c.Set(vstar.Property{Name: transpProp, Value: string(v)})
	hashing.SetXVSTAR(c)
}

File: rs/src/helpers/classification.rs
// SPDX-License-Identifier: MIT

//! `CLASS` (RFC 5545 §3.8.1.3) and `TRANSP` (§3.8.2.7).
//!
//! Each has a plain getter reporting absence, an `_or_default` getter
//! applying the RFC default, and a setter. The default is applied by
//! the `_or_default` form and by nothing else.
//!
//! An absent `CLASS` reports `None` rather than `Some(Public)`, even
//! though the RFC assigns `PUBLIC` by default, for two reasons. First,
//! `Option` means "present and recognized" everywhere else in this
//! module; one getter returning `Some` for an absent property would
//! make the whole convention untrustworthy. Second, a caller that must
//! distinguish "explicitly PUBLIC" from "unset" — round-trip fidelity,
//! diffing, supersession projection — cannot recover the distinction
//! once a getter has folded it away, whereas a caller wanting the
//! effective value gets it from [`class_or_default`] at zero cost.

use crate::enums::{Class, Transp};
use crate::hashing;
use crate::model::{Component, Property};
use crate::CompType;

/// The `CLASS` wire name.
const CLASS: &str = "CLASS";

/// The `TRANSP` wire name.
const TRANSP: &str = "TRANSP";

/// Reports whether `t` admits a `CLASS` property per RFC 5545 §3.8.1.3.
fn class_applies(t: &CompType) -> bool {
    *t == CompType::EVENT || *t == CompType::TODO || *t == CompType::JOURNAL
}

/// The parsed `CLASS` of `c`, or `None` when the property is absent or
/// carries an unrecognized value.
///
/// Spelled `class_of`, not `class`, in every port: `class` is reserved
/// in TypeScript, Python and PHP, and in Rust it would shadow the
/// [`Class`] type.
pub fn class_of(c: &Component) -> Option<Class> {
    c.get(CLASS).and_then(|p| Class::parse(&p.value))
}

/// The effective `CLASS` of `c`, applying the RFC 5545 §3.8.1.3 default
/// of `PUBLIC` when the property is absent or unrecognized.
///
/// An unrecognized value falls back to the default rather than being
/// surfaced: the RFC treats an unknown token as equivalent to `PRIVATE`
/// only for IANA / `X-` registrations it cannot resolve, and V\* does
/// not model those. Flagging the value is the validate layer's job, not
/// this accessor's.
pub fn class_or_default(c: &Component) -> Class {
    class_of(c).unwrap_or(Class::Public)
}

/// Writes `CLASS` and refreshes `X-VSTAR-HASH` last. No-op when `c`'s
/// type does not admit `CLASS`.
pub fn set_class(c: &mut Component, v: Class) {
    if !class_applies(&c.r#type) {
        return;
    }
    c.set(Property::new(CLASS, v.as_str()));
    hashing::set_x_vstar(c);
}

/// The parsed `TRANSP` of `c`, or `None` when absent or unrecognized.
/// See the module docs for why absence is not reported as the RFC
/// default here.
pub fn transp(c: &Component) -> Option<Transp> {
    c.get(TRANSP).and_then(|p| Transp::parse(&p.value))
}

/// The effective `TRANSP` of `c`, applying the RFC 5545 §3.8.2.7
/// default of `OPAQUE` — the event consumes free/busy time — when the
/// property is absent or unrecognized.
pub fn transp_or_default(c: &Component) -> Transp {
    transp(c).unwrap_or(Transp::Opaque)
}

/// Writes `TRANSP` and refreshes `X-VSTAR-HASH` last. No-op when `c` is
/// not a VEVENT: `TRANSP` is VEVENT-only per RFC 5545 §3.8.2.7.
pub fn set_transp(c: &mut Component, v: Transp) {
    if c.r#type != CompType::EVENT {
        return;
    }
    c.set(Property::new(TRANSP, v.as_str()));
    hashing::set_x_vstar(c);
}
